crates/hir-def/src/generics.rs

```diff
@@ -154,12 +155,28 @@ impl GenericParams {
         def: GenericDefId,
     ) -> Interned<GenericParams> {
         let _p = profile::span("generic_params_query");
+
+        let krate = def.module(db).krate;
+        let cfg_options = db.crate_graph();
+        let cfg_options = &cfg_options[krate].cfg_options;
+        let enabled_params = |params: &GenericParams, item_tree: &ItemTree| {
+            let enabled = |param| item_tree.attrs(db, krate, param).is_cfg_enabled(cfg_options);
+            Interned::new(GenericParams {
+                type_or_consts: (params.type_or_consts.iter())
+                    .filter_map(|(idx, param)| enabled(idx.into()).then(|| param.clone()))
+                    .collect(),
+                lifetimes: (params.lifetimes.iter())
+                    .filter_map(|(idx, param)| enabled(idx.into()).then(|| param.clone()))
+                    .collect(),
+                where_predicates: params.where_predicates.clone(),
+            })
```

Comment:
A bit unfortunate that we have to recollect everything even though the common case would be a no-op. It might make sense for us to iterate the params twice here, an initial pass to check if any are disabled, if so, do the collect dance, if not just clone the params. iterating twice should be pretty cheap since usually there aren't a lot of parameters anyways (also a comment explaining why we do this would be nice)

Reply:
Is `Interned::new()` also expensive or is it mainly the `.collect()`s? We could reuse the original `Interned<GenericParams>` if nothing is disabled

Reply:
`Interned` is just an `Arc` that does a global lookup on `Interned::new` to check if we can deduplicate the result, so if nothing is changed we'd just want to clone that as the code was doing before (as that is just a ref count increase)